### PR TITLE
[WIP] Add a Docker registry/pull-through cache

### PR DESCRIPTION
**Pull request by soulshake**
## What is the problem that this PR is trying to fix?

The purpose of this PR is to reduce egress for Docker Hub.

## What approach did you choose and why?

This PR:
- adds a GCE instance to serve as a Docker read-only registry/pull-through cache, and
- sets the `registry-mirrors` to the above instance on the Docker config of the default GCE build image

Addresses https://github.com/travis-ci/team-blue/issues/858

### CODE REVIEW

File: gce-staging-registry-1/main.tf [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+variable "gce_gcloud_zone" {}
+
+variable "index" {
+  default = 1
+}
+
+variable "project" {
+  default = "eco-emissary-99515"
```
Comment (meatballhat) [inline gce-staging-registry-1/main.tf]: I expect this should be `travis-staging-1`, yes?

File: gce-staging-registry-1/main.tf [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+variable "syslog_address_com" {}
+variable "syslog_address_org" {}
+
+variable "travisci_net_external_zone_id" {
+  default = "Z2RI61YP4UWSIO"
```
Comment (meatballhat) [inline gce-staging-registry-1/main.tf]: It's a trivial thing, but I would like to get away from referencing this directly and instead looking it up as with https://github.com/travis-ci/terraform-config/blob/9ee6c353d33d1589229c719db74f2a41761a6fe0/dns-production-0/main.tf#L45-L47

File: gce-staging-registry-1/main.tf [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    dynamodb_table = "travis-terraform-state"
+  }
+}
+
+provider "google" {
+  project = "${var.project}"
+  region  = "us-central1"
+}
+
+data "terraform_remote_state" "registry_staging_1" {
```
Comment (meatballhat) [inline gce-staging-registry-1/main.tf]: I don't see this being used, and I don't expect you will need to pull in the remote state of the same graph that is being built here.

File: modules/gce_docker_registry_cache/cloud-config.yml.tpl [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+  encoding: b64
+  path: /etc/docker/registry/config.yml
+- content: '${base64encode(file("${here}/cloud-init.bash"))}'
+  encoding: b64
+  path: /var/lib/cloud/scripts/per-instance/99-travis-cloud-init
+  permissions: '0750'
+- content: '${base64encode(syslog_address)}'
+  encoding: b64
+  path: /var/tmp/travis-run.d/syslog-address
+- content: '${base64encode(file("${here}/daemon-direct-lvm.json"))}'
```
Comment (meatballhat) [inline modules/gce_docker_registry_cache/cloud-config.yml.tpl]: Is direct-lvm storage helpful in this case? My understanding was that all blob storage was external.

File: modules/gce_docker_registry_cache/cloud-init.bash [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+  : "${VARTMP:=/var/tmp}"
+
+  export DEBIAN_FRONTEND=noninteractive
+
+  eval "$(tfw printenv docker)"
+  __setup_docker
+  __install_registry
+}
+
+__setup_docker() {
```
Comment (meatballhat) [inline modules/gce_docker_registry_cache/cloud-init.bash]: I think that this is now redundant/outdated given what `tfw` provides by default. See the squignix setup for a more recent example https://github.com/travis-ci/terraform-config/blob/master/modules/gce_squignix/cloud-init.bash

File: modules/gce_docker_registry_cache/main.tf [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+}
+
+data "template_file" "cloud_config" {
+  template = "${file("${path.module}/cloud-config.yml.tpl")}"
+
+  vars {
+    here = "${path.module}"
+
+    docker_registry_config = <<EOF
+${file("${path.module}/registry-config.yml")}
```
Comment (meatballhat) [inline modules/gce_docker_registry_cache/main.tf]: With this as the body, I would move the `${file(...)}` directly into the template.

File: modules/gce_remote_docker/daemon-direct-lvm.json [hunk 7]
```diff
@@ -17,5 +17,6 @@
   "tlsverify": true,
   "tlscacert": "/etc/docker/ca.pem",
   "tlscert": "/etc/docker/server-cert.pem",
-  "tlskey": "/etc/docker/server-key.pem"
+  "tlskey": "/etc/docker/server-key.pem",
+  "registry-mirrors": ["staging-1-docker-registry.gce-us-central1.travisci.net"]
```
Comment (meatballhat) [inline modules/gce_remote_docker/daemon-direct-lvm.json]: I assume this should not be hardcoded :grin:  Maybe the generated config in `tfw` should account for registry mirrors: https://github.com/travis-ci/tfw/blob/bfbffd80e1ddbcfd13cf58b9fbe435949e810d3a/bin/tfw#L1270-L1288

Comment (soulshake) [reply modules/gce_remote_docker/daemon-direct-lvm.json]: Ahhh I was looking all over for this. Thanks

File: modules/gce_docker_registry_cache/cloud-init.bash (hunk 5 above)
Comment (soulshake) [reply modules/gce_docker_registry_cache/cloud-init.bash]: Where is the docker config being passed to the docker daemon?

**Comment by soulshake** [comment]
Planning to just use [Container Registry's Docker Hub mirror](https://cloud.google.com/container-registry/docs/using-dockerhub-mirroring).

**Pull request closed without merging**